### PR TITLE
test: Ensure router's HTTP error messages are stable

### PR DESCRIPTION
**Pull request by shepmaster**
If you change the text of an error, the tests will fail.

If you add a new error variant to the `Error` enum but don't add it to the test, test compilation will fail with a "non-exhaustive patterns" message.

If you remove an error variant, test compilation will fail with a "no variant named `RemovedError`" message.

You can get the list of error variants and their current text via `cargo test -p router -- print_out_error_text --nocapture`.

A step towards accomplishing #5863

Co-authored-by: Carol (Nichols || Goulding) <carol.nichols@gmail.com>
Co-authored-by: Jake Goulding <jake.goulding@integer32.com>

### CODE REVIEW

**Comment by carols10cents** [comment]
I was kind of intrigued by the challenges of this issue, and I nerdsniped @shepmaster with it too :)

**Comment by shepmaster** [comment]
While @carols10cents and I were working on this, we noticed that some of the current error messages are... _suboptimal_.  Should we change some of these before documenting / committing to them? It doesn't necessarily have to be a part of this PR, but it could be.

/cc @rickspencer3

File: router/src/server/http.rs [hunk 1]
```diff
@@ -1307,4 +1310,143 @@ mod tests {
...
+                    #[allow(unreachable_patterns)]
+                    match dummy_error {
+                        $(Error::$variant { .. } => {},)*
+                    }
+                }
+            }
+
+            #[tokio::test]
+            async fn error_messages_match() {
+                $(
```
Comment (domodwyer) [inline router/src/server/http.rs:1333]:
```suggestion
            // A test that covers all errors given to this macro.
            async fn error_messages_match() {
                // Generate an assert for each error given to this macro.
                $(
```

File: router/src/server/http.rs [hunk 2]
```diff
@@ -1307,4 +1310,143 @@ mod tests {
...
+                    let e = Error::$variant $(($data))?;
+                    assert_eq!(e.to_string(), $msg);
+                )*
+            }
+
+            #[test]
+            fn print_out_error_text() {
```
Comment (domodwyer) [inline router/src/server/http.rs:1340]: Is this actually necessary? Wouldn't the failed asserts above spit out all the necessary strings?

**Review by domodwyer — COMMENTED** [review]
Code is 👌 just some readability / comment suggestions.

Seems to have missed some important inner-error variants that are worth filling it - they happen to be the most likely errors a user will see!

I am totally cool with merging this and having fixture tests (helps avoid bad errors!) but I'm not sure we should commit to stable error messages this early on.

File: router/src/server/http.rs (hunk 2 above)
Comment (carols10cents) [reply router/src/server/http.rs:1340]: This is actually for the case when all the asserts *pass* and we want to generate all the error messages to be transferred to some other documentation. If this doesn't sound useful, I'm happy to remove it!

**Review by domodwyer — APPROVED** [review]
Beautiful! Thanks for this 🙏

**Comment by domodwyer** [comment]
Merging this so it doesn't get conflicted :+1:

**Pull request merged**